modules/standard/IO.chpl

```diff
@@ -5663,14 +5769,332 @@ proc _channel.readLine(ref b: bytes,
+    `maxSize` bytes. The fileReader position is not moved.
+  :throws SystemError: Thrown if data could not be read from the ``fileReader``.
+*/
+proc fileReader.readTo(separator: bytes, ref b: bytes, maxSize=-1): bool throws {
+  var atEof = false;
+  on this._home {
+    try this.lock(); defer { this.unlock(); }
+
+    const (searchErr, _, bytesOffset) = _findSeparator(separator, maxSize, this._channel_internal);
+    if searchErr != 0 && searchErr != EEOF then try this._ch_ioerror(searchErr, "in fileReader.readTo(bytes)");
+    atEof = searchErr == EEOF && bytesOffset == 0;
+
+    const err = readStringBytesData(b, this._channel_internal, bytesOffset, 0);
+    if err then try this._ch_ioerror(err, "in fileReader.readTo(bytes)");
+  }
+  return !atEof;
+}
+
+/* helper for: readTo, readThrough, advanceThrough, advanceTo
+
+  looks for a sequence of bytes matching 'separator' in the
+  next 'maxSize' bytes in the channel
+
+  does not move the channel's pointer
+
+ returns: (0, true, byte_offset) if found
+          (EFORMAT, false, maxSize) if not found
+          (EFORMAT, false, 0) if separator is empty
+          (EEOF, false, bytes_to_eof) if EOF
+          (error_code, false, 0) system error
+*/
+private proc _findSeparator(separator: ?t, maxSize=-1, ch_internal): (errorCode, bool, int)
+  where t==string || t==bytes
+{
+  if separator.isEmpty() then return (EFORMAT:errorCode, false, 0);
+
+  const maxToRead = if maxSize < 0 then max(int) else maxSize,
+        sepLocal = separator.localize(),
+        numSepBytes = sepLocal.numBytes,
+        firstByte = sepLocal.byte(0);
+
+  var nextByte: int,
+      err: errorCode = 0,
+      foundSeparator = false,
+      numMatched = 0;
+
+  qio_channel_mark(false, ch_internal); // A
+  while true {
+    // advance to the the first byte in the separator
+    //  (separator's first byte is intentionally not consumed here
+    //   so that reverting B puts the pointer **before** the separator)
+    err = qio_channel_advance_past_byte(false, ch_internal, firstByte, /* consume */ false);
+    if err == EEOF {
+      break;
+    } else if err {
+      qio_channel_revert_unlocked(ch_internal); // A
+      return (err, false, 0);
+    }
+
+    // try to match the entire separator
+    qio_channel_mark(false, ch_internal); // B
+    numMatched = 0;
+    for i in 0..<numSepBytes {
+      nextByte = qio_channel_read_byte(false, ch_internal);
+      err = -nextByte;
+
+      if err == EEOF {
+        break;
+      } else if nextByte < 0 {
+        qio_channel_revert_unlocked(ch_internal); // B
+        qio_channel_revert_unlocked(ch_internal); // A
+        return (err, false, 0);
+      }
+
+      if nextByte == sepLocal.byte(i)
+        then numMatched += 1;
+        else break;
+    }
+    qio_channel_revert_unlocked(ch_internal); // B
+
+    if numMatched == numSepBytes {
+      foundSeparator = true;
+      break;
+    }
+
+    // consume an additional byte
```

Comment:
I'm not following why it needs to consume an additional byte. Anyway if you keep it could you put an `if nextByte < 0 { ... }` as above.